amqp/connection.go

```diff
@@ -2,52 +2,163 @@ package amqp
 
 import (
 	"errors"
-	"github.com/eventials/goevents/messaging"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
 	amqplib "github.com/streadway/amqp"
+
+	"github.com/eventials/goevents/messaging"
 )
 
+// Connection with an AMQP peer.
 type Connection struct {
+	config     ConnectionConfig
+	m          sync.Mutex
+	url        string
 	connection *amqplib.Connection
+	closed     bool
+
+	reestablishs []chan bool
+}
+
+// ConnectionConfig to be used when creating a new connection.
+type ConnectionConfig struct {
+	reconnectInterval time.Duration
+	reconnectRetries  int
 }
 
 // NewConnection returns an AMQP Connection.
+// Uses a default ConnectionConfig with 2 second of reconnect interval and 10 retries.
 func NewConnection(url string) (messaging.Connection, error) {
+	return NewConnectionWithConfig(url, ConnectionConfig{
+		reconnectInterval: 2 * time.Second,
+		reconnectRetries:  10,
+	})
+}
+
+// NewConnectionWithConfig returns an AMQP Connection.
+func NewConnectionWithConfig(url string, config ConnectionConfig) (messaging.Connection, error) {
 	conn, err := amqplib.Dial(url)
 
 	if err != nil {
 		return nil, err
 	}
 
-	return &Connection{
-		conn,
-	}, nil
+	connection := &Connection{
+		url:        url,
+		connection: conn,
+		config:     config,
+	}
+
+	go connection.handleConnectionClose()
+
+	return connection, nil
 }
 
+// NotifyConnectionClose returns a channel to notify when the connection closes.
 func (c *Connection) NotifyConnectionClose() <-chan error {
 	ch := make(chan error)
 
 	go func() {
-		ch <- errors.New((<-c.connection.NotifyClose(make(chan *amqplib.Error))).Error())
+		se := <-c.connection.NotifyClose(make(chan *amqplib.Error))
+
+		ch <- errors.New(se.Error())
```

Comment:
Yeah you're right after all. We did this initially because we thought that we couldn't pass *amqplib.Error to a error. Turns out it can, so fixing it...